README
LastMoid is Plasma applet designed for last.fm users.
It is able to display music charts for various periods of time and recently listened tracks.
Based on latest Plasma widgets API with straightforward configuration.

Links :
http://github.com/vavrusa/lastmoid
http://www.last.fm/home
http://en.wikipedia.org/wiki/Last.fm

Authors:
Marek Vavrusa <marek@vavrusa.com>
Damien Lévin <dml_aon@hotmail.com>

Installation
------------

tar xvzf lastmoid-0.6.5.tar.gz
cd lastmoid-0.6.5
mkdir build && cd build
cmake -DCMAKE_INSTALL_PREFIX=`kde4-config --prefix` ..
make
make install (as root or with sudo)



ChangeLog
0.6.5
    - Elided text for labels
    - Tooltip containing full text
    - Code cleanup and comments (more to follow)
    - Reflects invalid (missing) configuration
    - Animated bars appearance/change
    - Intelligent list management
    - Fixed weekly charts data
0.6
    - Complete code refactoring and cleanup
    - Unified header handling, picture loading
    - Refactored configuration
    - Removed old rendering code
    - Moved to Plasma-based API
    - Reflecting busy state
0.5
    - Cut too long data
    - Recent played tacks available

0.4
    - Refresh bug fixed
    - 3 months / 12 months / Overall charts available
    - Album charts available
    - Add an Icon

0.3
    - Bug fixes
    - New logo
    - Adding a tool to set :
      * LastFm user
      * Chart

0.1
    - First draft of the weekly artist chart


track.h
/***************************************************************************
 *   Copyright (C) 2009 by Marek Vavrusa <marek@vavrusa.com>      	   *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 3 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#ifndef TRACK_H
#define TRACK_H

#include <Plasma/Label>
#include <QFlags>
class QGraphicsWidget;

/** Links track data with interactive label.
  */
class Track : public Plasma::Label
{
   Q_OBJECT
   Q_PROPERTY(float bar READ barValue WRITE setBarValue)
   Q_PROPERTY(QColor barColor READ barColor WRITE setBarColor)

   public:

   /** Flags */
   enum Flag {
      ElideText  = 0x01,
      EdgeMark   = 0x02,
      NowPlaying = 0x04
   };
   Q_DECLARE_FLAGS(Flags, Flag)

   /** Format */
   enum Attribute {
      Name       = 0x01, /// Track name
      Artist     = 0x02, /// Track artist
      Album      = 0x04, /// Track album
      Date       = 0x08, /// Last time played
      PlayCount  = 0x10  /// Play count

   };
   Q_DECLARE_FLAGS(Format, Attribute)

   /** Constructor */
   explicit Track(QGraphicsWidget *parent = 0);

   /**  Return track attribute.
     * \param key
     * \return track attribute
     */
   const QString& attrib(Attribute key) {
      return mData[key];
   }

   /** Set track attribute to given value
     * \param key
     * \param val - new value
     * \return newly assigned value
     */
   const QString& setAttribute(Attribute key, const QString& val) {
      return mData[key] = val;
   }

   /** Display format.
     * Usage:
     * %n - track name
     * %l - album
     * %a - artist
     * %d - date played
     * %p - play count
     * \return display format
     */
   const QString& format() {
      return mFormat;
   }

   /** Set display format.
     * Usage:
     * %n - track name
     * %l - album
     * %a - artist
     * %d - date played
     * %p - play count
     * \param fmt - format string
     */
   void setFormat(const QString& fmt);

   /** Return bar flags.
       \return flags
     */
   Flags flags() { return mFlags; }
   
   /** Set bar flags.
       \param flags
     */
   void setFlags(Flags flags);

   /** Return bar percentage 0.0 - 1.0f;
       \param bar percentage
     */
   float barValue() { return mBarValue; }
   /** Set bar percentage to given value
       \param bar float value 0.0f - 1.0f
       \return previous value
     */
   float setBarValue(float bar);

   /** Return QString representation.
     */
   QString toString();

   /** Animate given property.
     * \param property - given Q_PROPERTY (opacity, pos, barValue ...)
     * \param from - starting value
     * \param to - end value
     */
   void animate(const QByteArray& property, const QVariant& from, const QVariant& to);

   /** Current bar color
       \return bar color
     */
   QColor barColor() { return mBarColor; }

   /** Set bar color to given value
       \param color new color
       \return previous color
     */
   QColor setBarColor(const QColor& color);

   protected:

   /** Update text and tooltip for widget size.
     */
   void updateLabel();
   void resizeEvent(QGraphicsSceneResizeEvent *event);
   void paint(QPainter *painter,
              const QStyleOptionGraphicsItem *option,
              QWidget *widget);

   private:  
   Flags mFlags;     /// Flags
   float mBarValue;  /// Progress 0.0f - 1.0f
   QColor mBarColor; /// Bar color (default red)
   QString mFormat;  /// Display text
   QMap<Attribute,QString> mData; /// Data map
};

Q_DECLARE_OPERATORS_FOR_FLAGS(Track::Flags)

#endif // TRACK_H


track.cpp
/***************************************************************************
 *   Copyright (C) 2009 by Marek Vavrusa <marek@vavrusa.com>      	   *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 3 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/


#include <QtGlobal>
#include <Plasma/Animator>
#include <QGraphicsSceneResizeEvent>
#include <QGraphicsWidget>
#include <QFontMetrics>
#include <QPainter>

// Animation
#if QT_VERSION >= 0x040600
#include <QPropertyAnimation>
#endif

#include "track.h"

Track::Track(QGraphicsWidget *parent) :
    Plasma::Label(parent), mBarValue(0.0), mBarColor(QColor(215, 0, 25, 96))
{
   // Set default size
   QFontMetrics fnm(font());
   setScaledContents(false);
   setSizePolicy(QSizePolicy::MinimumExpanding, QSizePolicy::Fixed);
   setMaximumHeight(fnm.height() * 1.25);
}

void Track::setFlags(Flags flags)
{
   mFlags = flags;
   updateLabel();
}

void Track::setFormat(const QString &fmt)
{
   mFormat = fmt;
   updateLabel();
}

void Track::animate(const QByteArray& property, const QVariant& from,
                    const QVariant& to)
{
// At least Qt 4.6.0 with QtKinetic
#if QT_VERSION >= 0x040600
    // Animate progress
    QPropertyAnimation* anim = new QPropertyAnimation(this, property, this);
    anim->setTargetObject(this);
    anim->setPropertyName(property);
    anim->setDuration(500);
    anim->setStartValue(from);
    anim->setEndValue(to);
    anim->start(QPropertyAnimation::DeleteWhenStopped);
#else
    // Set progress instantly
    setProperty(property, to);
    Q_UNUSED(from);
#endif
}

float Track::setBarValue(float val)
{
    float oldVal = mBarValue;
    mBarValue = val;
    update();

    return oldVal;
}

QColor Track::setBarColor(const QColor& color)
{
   QColor oldColor = mBarColor;
   mBarColor = color;
   update(boundingRect());
   return oldColor;
}

void Track::updateLabel()
{
   // Get text
   QString text = toString();

   // Update toolTip
   if(!attrib(PlayCount).isEmpty())
    setToolTip(text + tr(" (%1 plays)").arg(attrib(PlayCount)));
   else
    setToolTip(text);

   // Elide text
   if(flags() & ElideText) {
      QFontMetrics fnm(font());
      setText(fnm.elidedText(text, Qt::ElideRight, contentsRect().width()));
   }
}

QString Track::toString() {
   QString res(format());
   res.replace("%n", attrib(Name));
   res.replace("%l", attrib(Album));
   res.replace("%a", attrib(Artist));   
   res.replace("%d", attrib(Date));
   res.replace("%p", attrib(PlayCount));
   return res;
}

void Track::resizeEvent(QGraphicsSceneResizeEvent *event)
{
   Plasma::Label::resizeEvent(event);

   // Update label
   updateLabel();
}

void Track::paint(QPainter *painter,
           const QStyleOptionGraphicsItem *option,
           QWidget *widget)
{
   // Paint background
   if(mBarValue > 0.0) {
      painter->setRenderHint(QPainter::Antialiasing);

      QRectF barRect = boundingRect();
      barRect.setWidth(barRect.width() * mBarValue);

      // Bar color
      QColor barColor(mBarColor);
      if(flags() & NowPlaying)
         barColor = QColor(255, 252, 202, 96); // Soft yellow

      // Edge decoration for progress
      if(flags() & EdgeMark) {
         float edgeLen = barRect.height() * 0.5;
         barRect.setWidth(barRect.width()  - edgeLen);
         QPainterPath path(barRect.topLeft());
         path.lineTo(barRect.topRight());
         path.lineTo(barRect.topRight() + QPointF(edgeLen, barRect.height() * 0.5));
         path.lineTo(barRect.bottomRight());
         path.lineTo(barRect.bottomLeft());
         path.closeSubpath();
         painter->fillPath(path, QBrush(barColor));
      }
      else {
         painter->fillRect(barRect, QBrush(barColor));
      }
   }

   // Paint widget
   Plasma::Label::paint(painter, option, widget);
}

#include "track.moc"


barlabel.h
/***************************************************************************
 *   Copyright (C) 2009 by Marek Vavrusa <marek@vavrusa.com>      	   *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 3 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#ifndef BARLABEL_H
#define BARLABEL_H

#include <Plasma/Label>
class QGraphicsWidget;

class Track : public Plasma::Label
{
   Q_OBJECT
   Q_PROPERTY(int textFlags READ textFlags WRITE setTextFlags)
   Q_PROPERTY(float bar READ barValue WRITE setBarValue)
   Q_PROPERTY(QColor barColor READ barColor WRITE setBarColor)

   public:
   explicit BarLabel(QGraphicsWidget *parent = 0);

   /** Overloaded for storing original text.
       Stored text used for elided.
       \param text text string
     */
   void setText(const QString& text) {
      Plasma::Label::setText(mText = text);
   }

   /** Return bitmask of text flags defined in BarLabel::Flags
       \return bitmask of flags
     */
   int textFlags() { return mTextFlags; }
   /** Set text flags defined in BarLabel::Flags
       \param flags bitmask of flags
     */
   void setTextFlags(int flags);

   /** Return bar percentage 0.0 - 1.0f;
       \param bar percentage
     */
   float barValue() { return mBarValue; }
   /** Set bar percentage to given value
       \param bar float value 0.0f - 1.0f
       \return previous value
     */
   float setBarValue(float bar);

   /** Animate given property.
     */
   void animate(const QByteArray& property, const QVariant& from, const QVariant& to);

   /** Current bar color
       \return bar color
     */
   QColor barColor() { return mBarColor; }

   /** Set bar color to given value
       \param color new color
       \return previous color
     */
   QColor setBarColor(const QColor& color);

   /** Text flags */
   enum TextFlags {
      NoFlags   = 0x00,
      ElideText = 0x01,
      EdgeMark  = 0x02,
      Playing   = 0x04
   };

   protected:
   void resizeEvent(QGraphicsSceneResizeEvent *event);
   void paint(QPainter *painter,
              const QStyleOptionGraphicsItem *option,
              QWidget *widget);

   private:
   QString mText;   // Original text
   QColor mBarColor;// Bar color (default red)
   float mBarValue; // Progress 0.0f - 1.0f
   int mTextFlags;  // Text flags
};

#endif // BARLABEL_H


lastmoid.h
/***************************************************************************
 *   Copyright (C) 2008 by Damien Lévin <dml_aon@hotmail.com>      	   *
 *                 2009 by Marek Vavrusa <marek@vavrusa.com>      	   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 3 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/
 
#ifndef LASTMOID_H
#define LASTMOID_H

#include <Plasma/Applet>
#include <QByteArray>
#include <QUrl>
#include <KConfigGroup>

class Lastmoid : public Plasma::Applet
{
    Q_OBJECT
    public:

        Lastmoid(QObject *parent, const QVariantList &args);
        ~Lastmoid();

        void paintInterface(QPainter *painter, const QStyleOptionGraphicsItem *option,const QRect& contentsRect);
        void fetch();
        void loadConfig();

    public slots:
        void init();
        void refresh();
        void httpQuery(const QUrl& url);
        void httpResponse(int id, bool error);

    protected slots:
        void configAccepted();
        void setBusy(bool val);

    protected:
        bool parseUserData(const QByteArray& data);
        bool parseStatData(const QByteArray& data);
        bool parseRecentTracks(const QByteArray& data);
        void createConfigurationInterface(KConfigDialog *parent);

    private:
        struct Private;
        Private *d;
};

K_EXPORT_PLASMA_APPLET(lastmoid, Lastmoid)
#endif


lastmoid.cpp
/***************************************************************************
 *   Copyright (C) 2008 by Damien Lévin <dml_aon@hotmail.com>      	   *
 *                 2009 by Marek Vavrusa <marek@vavrusa.com>      	   *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 3 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#include <QGraphicsLinearLayout>
#include <QDomDocument>
#include <QDomElement>
#include <QPainter>
#include <QHttp>
#include <QTimer>
#include <QList>
#include <Plasma/Theme>
#include <Plasma/ScrollWidget>
#include <Plasma/BusyWidget>
#include <KConfigDialog>
#include "lastmoid.h"
#include "track.h"
#include "ui_lastmoidConfig.h"
#include <QFile>
// Data type (mapped to selector)
enum Data {
   RecentTracks = 0,
   TopAlbums    = 1,
   TopArtists   = 2,
   TopTracks    = 3
};

// Data period
enum Period {
   Weekly       = 0,
   Overall      = 1,
   ThreeMonths  = 2,
   SixMonths    = 3,
   TwelveMonths = 4
};

// State tracking
enum State {
   NotFound    = 0x00, // User not found
   Finalizing  = 0x01, // Querying user
   Identified  = 0x02  // Matching user found
};

struct Lastmoid::Private
{
   Private() : interval(0), data(0), period(0), limit(0), lastDate(0), connId(-1),
               state(NotFound), layout(0), dataLayout(0), scrollWidget(0),
               busyWidget(0)
   {}

   // Config
   int interval;
   int data;
   int period;
   int limit;
   QString dataStr;
   QString periodStr;
   QString login;

   // Containers
   int lastDate;
   int connId;
   State state;
   QImage avatar;
   QHttp http;
   QTimer timer;
   Ui::lastmoidConfig configUi;
   Plasma::Svg svgLogo;

   // Widgets
   QList<Track*> trackList;
   QGraphicsLinearLayout* layout;
   QGraphicsLinearLayout* dataLayout;
   Plasma::ScrollWidget* scrollWidget;
   Plasma::BusyWidget*   busyWidget;
};

Lastmoid::Lastmoid(QObject *parent, const QVariantList &args)
   : Plasma::Applet(parent, args), d(new Private)
{
   // Init widget defaults
   d->svgLogo.setImagePath("widgets/lastmoid");
   setBackgroundHints(DefaultBackground);
   resize(220, 300);
}


Lastmoid::~Lastmoid()
{
   // Stop timer
   d->timer.stop();
   delete d;
}

void Lastmoid::init()
{
   // Prepare scroll widget
   QFontMetrics fm(font());
   d->scrollWidget = new Plasma::ScrollWidget(this);
   QGraphicsWidget* dataWidget = new QGraphicsWidget(d->scrollWidget);
   d->scrollWidget->setWidget(dataWidget);
   d->dataLayout = new QGraphicsLinearLayout(Qt::Vertical, dataWidget);
   d->busyWidget = new Plasma::BusyWidget(this);
   d->busyWidget->hide();

   // Contents widget
   d->layout = new QGraphicsLinearLayout(Qt::Vertical, this);
   d->layout->setContentsMargins(0,60 + fm.height() * 0.5,0,0);

   // Connect timer and http results
   connect(&d->timer, SIGNAL(timeout()), this, SLOT(refresh()));
   connect(&d->http, SIGNAL(requestFinished(int, bool)),this, SLOT(httpResponse(int, bool)));

   // Load config and start
   loadConfig();
   if(!d->login.isEmpty()) {
      fetch();
   }
}

void Lastmoid::refresh()
{
   // Reset timer
   d->timer.stop();
   d->timer.setInterval(d->interval * 60 * 1000);
   d->timer.start();

   // Fetch immediately
   fetch();
}

void Lastmoid::createConfigurationInterface(KConfigDialog *parent)
{
   // Create configuration widget
   QWidget *widgetConfig = new QWidget;
   KConfigGroup configGroup = config();
   d->configUi.setupUi(widgetConfig);
   d->configUi.user->setText(configGroup.readEntry("user"));
   d->configUi.dataType->setCurrentIndex(configGroup.readEntry("dataType", "0").toInt());
   d->configUi.dataPeriod->setCurrentIndex(configGroup.readEntry("dataPeriod", "1").toInt());
   d->configUi.timer->setValue(configGroup.readEntry("timer", "5").toInt());
   d->configUi.limit->setValue(configGroup.readEntry("limit", "20").toInt());

   // Connect controls
   connect(parent, SIGNAL(okClicked()), this, SLOT(configAccepted()));
   parent->setButtons(KDialog::Ok | KDialog::Cancel);
   parent->addPage(widgetConfig, i18n("Configuration"), icon());
}

void Lastmoid::configAccepted()
{
   // Save configuration
   KConfigGroup configGroup = config();
   configGroup.writeEntry("user", d->configUi.user->text());
   configGroup.writeEntry("dataType", QString::number(d->configUi.dataType->currentIndex()));
   configGroup.writeEntry("dataPeriod", QString::number(d->configUi.dataPeriod->currentIndex()));
   configGroup.writeEntry("timer", QString::number(d->configUi.timer->value()));
   configGroup.writeEntry("limit", QString::number(d->configUi.limit->value()));
   configGroup.sync();

   // Reload configuration
   loadConfig();

   // Remove all elements
   while(!d->trackList.empty()) {
      d->dataLayout->removeAt(0);
      d->trackList.front()->deleteLater();
      d->trackList.pop_front();
   }

   // Reset last date and timers
   d->lastDate = 0;
   d->timer.stop();

   // Reset avatar and reset state
   d->avatar = QImage();
   d->state = NotFound;

   // Refresh if user available
   if(!d->login.isEmpty()) {
      d->scrollWidget->show();
      refresh();
   }
   else {
      d->scrollWidget->hide();
   }
}


void Lastmoid::loadConfig()
{
   // Translate index to QString (watch indexes in lastmoidConfig.ui)
   static QString dataTable[4] = {
      "recentTracks", "album", "artist", "track" };
   static QString periodTable[5] = {
      "7day", "overall", "3month", "6month", "12month" };

   // Load configuration data
   KConfigGroup configGroup = config();
   d->login = configGroup.readEntry("user");
   d->data = configGroup.readEntry("dataType", "0").toInt();\
   d->dataStr = dataTable[d->data % 4];
   d->period = configGroup.readEntry("dataPeriod", "0").toInt();
   d->periodStr = periodTable[d->period % 5];
   d->interval =  QString(configGroup.readEntry("timer", "5")).toInt();
   d->limit =  QString(configGroup.readEntry("limit", "20")).toInt();

   // Interval range check
   if(d->interval <= 0)
      d->interval = 5;

   // Update scene rect
   update();
}

void Lastmoid::fetch()
{
   // Is user initialised?
   QString url("http://ws.audioscrobbler.com/2.0/?method=");
   switch(d->state) {
   case Identified:

      // Recent Tracks
      if(d->data == RecentTracks) {
         url.append("user.getrecenttracks&user=" + d->login + "&limit=" + QString::number(d->limit));
      }
      else {
         // Charts
         url.append("user.gettop" + d->dataStr + "s&user=" + d->login + "&period=" + d->periodStr);
      }
      break;

   default:
      url.append("user.getinfo&user=" + d->login);
      break;
   }

   // Execute request
   httpQuery(QUrl(url + "&api_key=b6eb61f91b89e55548dd14732ee0b8a1"));
}

void Lastmoid::httpQuery(const QUrl &url)
{
   d->http.setHost(url.host());
   d->connId = d->http.get(url.toString());
   //qDebug() << "Query (" << d->connId << "): " << url.toString();
}

void Lastmoid::httpResponse(int id, bool error)
{
   // Invalid request
   if(d->connId != id) {
      return;
   }

   //qDebug() << "Response (" << id << "): received " << d->http.bytesAvailable();

   // Error checking
   if(error) {
      qWarning("received error during HTTP fetch.");
      update();
      return;
   }

   // Evaluate result
   QByteArray data (d->http.readAll());
   bool result = false;
   switch(d->state) {
   case NotFound: // Idle state
      setBusy(true);
      result = parseUserData(data);
      break;

   case Finalizing: // Awaiting user data result

      // Create avatar
      d->avatar.loadFromData(data, "JPG");
      if(!d->avatar.isNull()) {
         d->avatar = d->avatar.scaledToHeight(60);
         d->state = Identified;
         result = true;
         refresh();
         update();
      }
      break;

   case Identified: // Awaiting data
      // Parse result
      if(d->data == RecentTracks) result = parseRecentTracks(data);
      else                        result = parseStatData(data);

      // Remove busy widget
      if(result && d->busyWidget->isVisible()) {
         setBusy(false);
      }
      break;

   default:
      break;
   }
}

bool Lastmoid::parseStatData(const QByteArray& data)
{
   // Create DOM and import data
   QDomDocument doc("?xml version=\"1.0\" encoding=\"utf-8\" ?");
   QDomElement root, element;
   doc.setContent(data);
   root = doc.firstChildElement("lfm");
   element = root.firstChildElement(d->dataStr);

   // Data period element
   element = root.firstChildElement("top" + d->dataStr + "s");

   // Enter group
   element = element.firstChildElement(d->dataStr);
   if(element.isNull()) {
      return false;
   }

   // Evaluate elements
   QFontMetrics fnm(font());
   QString artist, name;
   int maxCount = element.firstChildElement("playcount").text().toInt();
   for(int i = 0; i < d->limit && !element.isNull(); element = element.nextSiblingElement(d->dataStr)) {

      // Get basic data
      artist = element.firstChildElement("artist").firstChildElement("name").text();
      name = element.firstChildElement("name").text();

      // Check track list for match
      Track* track = 0;
      if(i < d->trackList.size()) {

         // Attempt to find track
         for(int k = 0; k < d->trackList.size(); ++k) {

            // Expect track here
            track = d->trackList.at(k);

            // Compare artist - name
            if(track->attrib(Track::Artist) == artist &&
               track->attrib(Track::Name) == name) {

               // Found it, move and save
               if(i != k) {
                  d->trackList.move(k, i);
                  d->dataLayout->removeItem(track);
                  d->dataLayout->insertItem(i, track);
               }
               break;
            }
         }
      }

      // Create or reuse label
      if(track == 0) {

         // Reuse on list exceeding limit
         if(d->trackList.size() == d->limit) {
            // Take last
            track = d->trackList.back();
            d->trackList.move(d->trackList.size() - 1, 0);

         } else {
            // Create new track record
            track = new Track(this);

            // Update flags
            track->setFlags(Track::EdgeMark|Track::ElideText);
            d->trackList.insert(i, track);
         }

         // Reinsert
         d->dataLayout->insertItem(i, track);

         // Update format
         track->setAttribute(Track::PlayCount, element.firstChildElement("playcount").text());
         switch(d->data) {
         case TopAlbums:
         case TopTracks:
            track->setAttribute(Track::Artist, element.firstChildElement("artist").firstChildElement("name").text());
            track->setAttribute(Track::Name, element.firstChildElement("name").text());
            track->setFormat(" %a - %n");
            break;
         case TopArtists:
            track->setAttribute(Track::Artist, element.firstChildElement("name").text());
            track->setFormat(" %a");
            break;
         default:
            break;
         }
      }

      // Update play count
      track->animate("bar", track->barValue(), track->attrib(Track::PlayCount).toInt() / (float) maxCount);
      ++i;
   }

   return true;
}

bool Lastmoid::parseRecentTracks(const QByteArray& data)
{
   // Create DOM and import data
   QDomDocument doc("?xml version=\"1.0\" encoding=\"utf-8\" ?");
   QDomElement root, element;

   doc.setContent (data);
   root = doc.firstChildElement("lfm");
   element = root.firstChildElement("recenttracks");
   element = element.lastChildElement("track"); // Reverse read

   // Check expected element
   if(element.isNull())
      return false;

   // Get new items
   static bool lastFlip = false;
   for(;!element.isNull(); element = element.previousSiblingElement("track")) {

      // Check last date
      int uts = element.firstChildElement("date").attribute("uts").toInt();

      // Check "Now Playing"
      if(uts == 0) {
         qDebug() << "fixme: track is now playing (skipping)";
         continue;
      }

      // Check if is newer
      if(uts <= d->lastDate)
         continue;

      // Update date
      d->lastDate = uts;

      // Create label
      Track* track = 0;

      // Create or reuse
      if(d->trackList.size() < d->limit) {
         // Create track
         track = new Track(this);
         track->setFlags(Track::ElideText);
         d->trackList.push_front(track);
      }
      else {
         // Take last
         track = d->trackList.back();
         d->trackList.move(d->trackList.size() - 1, 0);
      }

      // Reinsert
      d->dataLayout->insertItem(0, track);

      // Update attributes
      track->setAttribute(Track::Name, element.firstChildElement("name").text());
      track->setAttribute(Track::Artist, element.firstChildElement("artist").text());
      track->setFormat(" %a - %n");

      // Flip bar value
      lastFlip = !lastFlip;
      track->setBarValue(1.0 * (int) lastFlip);

      // Animate insert
      if(d->trackList.size() == d->limit)
         track->animate("opacity", 0.0, 1.0);
   }

   return true;
}

bool Lastmoid::parseUserData(const QByteArray& data)
{
   // Create DOM and import data
   QDomDocument doc("?xml version=\"1.0\" encoding=\"utf-8\" ?");
   QDomElement root, element;

   doc.setContent (data);
   root = doc.firstChildElement("lfm");
   element = root.firstChildElement("user");
   element = element.firstChildElement("image");

   // Check starting element
   if(!element.isNull()) {

      // Fetch avatar
      d->state = Finalizing;
      httpQuery(QUrl(element.text()));
      return true;
   }

   return false;
}

void Lastmoid::paintInterface(QPainter *p, const QStyleOptionGraphicsItem *option, const QRect &contentsRect)
{
   // Applet paint
   Plasma::Applet::paintInterface(p, option, contentsRect);

   // Measures
   QFont font = Plasma::Theme::defaultTheme()->font(Plasma::Theme::DefaultFont);
   QFontMetrics fm(font);
   int widgetWidth = (int)contentsRect.width();
   int hAlign = fm.height();

   // SVG logo
   p->setRenderHint(QPainter::SmoothPixmapTransform);
   p->setRenderHint(QPainter::Antialiasing);
   d->svgLogo.resize(94, 48);
   d->svgLogo.paint(p,hAlign, (int)contentsRect.top());


   p->save();

   // User string
   if(!d->login.isEmpty()) {
      QPoint headerPt(contentsRect.topLeft());
      headerPt.setX(hAlign);
      headerPt.setY(headerPt.y() + d->svgLogo.size().height() + 2);
      p->setPen(QColor(213,13,6));
      p->drawText(headerPt, d->login);
   } else {
      // Placeholder
      QRect msgRect(contentsRect);
      msgRect.setTop(msgRect.top() + d->svgLogo.size().height() + 2);
      msgRect.adjust(hAlign,hAlign,-hAlign,-hAlign);
      msgRect.setHeight(msgRect.height() * 0.5);

      // Icon
      KIcon icon("configure");
      p->setOpacity(0.8);
      icon.paint(p, msgRect, Qt::AlignTop|Qt::AlignHCenter);
      p->setOpacity(1.0);

      // Text
      msgRect.moveTop(msgRect.top() + msgRect.height());
      p->drawText(msgRect, Qt::AlignCenter, tr("User not set."));

   }

   // Avatar
   if(!d->avatar.isNull()) {

      // Create clip path
      QPainterPath avClip;
      QRect avFrame((int)contentsRect.left() + widgetWidth - 60, (int)contentsRect.top(), 60, 60);
      avClip.addRoundedRect(QRectF(avFrame), 10.0, 10.0);

      // Draw clipped avatar
      p->setClipPath(avClip);
      if(!d->avatar.isNull())
         p->drawImage(avFrame, d->avatar);
      p->setClipping(false);

      // Avatar border
      QPen pen(p->pen());
      pen.setColor(Qt::lightGray);
      pen.setWidth(2);
      p->setPen(pen);
      p->drawRoundedRect(avFrame, 10.0, 10.0);
   }

   p->restore();
}

void Lastmoid::setBusy(bool val)
{
   // Remove current widget
   if(d->layout->count())
      d->layout->removeAt(0);

   // If true, replace widget with "Busy" indicator
   if(val) {
      d->layout->addItem(d->busyWidget);
      d->busyWidget->show();
      d->scrollWidget->hide();
   }
   else {
      d->layout->addItem(d->scrollWidget);
      d->busyWidget->hide();
      d->scrollWidget->show();
   }
}

#include "lastmoid.moc"
